### PR TITLE
Cylc review update

### PR DESCRIPTION
**Pull request by datamel**
Closes https://github.com/cylc/cylc-flow/issues/5135

I managed to get SQLite3 traceback with a workflow where the `flow.cylc` had been deleted (for me, a cylc test battery test had not fully cleaned up correctly). This check was added in  https://github.com/cylc/cylc-flow/pull/5111. I think if someone deletes the suite.rc/flow.cylc files they should be able to continue to use cylc review without traceback, unless I am missing something?   

Investigation continues into `ProgrammingError: SQLite objects created in a thread can only be used in that same thread` and `CylcReviewService object is not callable` errors.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Tests are included (or explain why tests are not needed - legacy).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request not required.
- [ ] If this is a bug fix, PRs raised to both master and the relevant maintenance branch.

### CODE REVIEW

**Comment by datamel** [comment]
Update to this, there has been another bug found in cylc review. While viewing a workflow that had not yet been run, an old database file `cylc-suite.db` has been created. This renders the workflow unable to play at cylc8. This is also causing these workflows that have not run before to appear at the top of the list of running workflows, since they have been 'updated' recently.

**Comment by hjoliver** [comment]
Roger that, still holding off on the "final" cylc 7 release :grin:

**Comment by datamel** [comment]
Another Cylc review bug found today, workflows which have used the flow feature to run tasks multiple times are causing duplicate entries.

**Comment by datamel** [comment]
Update, last of the bugs fixed 
- problems with threading of database object, this resulted in a sqlite ProgrammingError. This is difficult to reproduce, no clear steps and was seemingly random but occurred fairly frequently on master
- some db _exec calls were not closed. 
- version files now should be visible
- this last one is undocumented as an issue, summary to reproduce here for reviewers:
Run a cylc 8 workflow and trigger an already run task with `--flow=new`
On `7.8.x` duplicate task entries are found, in this branch there should only be the correct number showing.

This is now ready for review.

File: lib/cylc/review_dao.py
Comment (datamel) [inline lib/cylc/review_dao.py:224]: This is imported down here due to a circular import problem. Unsure if there is a better way of solving this.

File: lib/cylc/review.py
Comment (wxtim) [inline lib/cylc/review.py:843]: Nice. Since it's a staticmethod couple of unit tests might be nice. You can probably mock the results of os.path.exists so that you don't have to touch the file system.

File: lib/cylc/review.py
Comment (wxtim) [inline lib/cylc/review.py:754]: Why not just have the method in the condition? Do you re-use `cylc_8`?

File: lib/cylc/review_dao.py
Comment (wxtim) [reply lib/cylc/review_dao.py:224]: Possible alternative - make `is_cylc8` a module function (not a class method) and using
`from cylc.review import is_cylc8`. Might look more elegant.

**Comment by wxtim** [comment]
I've done a code review, nothing that I'm going to demand you change, although I think some suggestions worth considering.

I would like to see a unit test of the `is_cylc8` function.

I'm going to do a practical review.

File: lib/cylc/review_dao.py
Comment (datamel) [reply lib/cylc/review_dao.py:224]: This import is also needed for the call to `_get_user_home`, not just for `is_cylc8` so making it a module function won't help with that unfortunately. I did look at moving this out but I don't think a big refactor is a good idea.

**Review by wxtim — APPROVED** [review]
I've had a fair poke around the Cylc codebase looking for files we've missed.

**Review by wxtim — CHANGES_REQUESTED** [review]
I've commented.

**Comment by datamel** [comment]
> I've commented.

Test added, responded to all your comments I think. I can't spot anything you have said following your second review where you have requested changes though, maybe github is being funny for me.

File: lib/cylc/tests/test_review.py
Comment (wxtim) [inline lib/cylc/tests/test_review.py]:
```suggestion
    assert CylcReviewService.is_cylc8(temp_workflow_dir) == expected

```

File: lib/cylc/tests/test_review.py
Comment (wxtim) [inline lib/cylc/tests/test_review.py:49]: Why not just tmp_path fixture?

File: lib/cylc/tests/test_review.py
Comment (wxtim) [inline lib/cylc/tests/test_review.py:52]: IMO also a case for `cylc_8 is None` - i.e. there is neither a suite.rc or a flow.cylc.

Comment (datamel) [reply lib/cylc/tests/test_review.py]: Added.

File: lib/cylc/tests/test_review.py
Comment (datamel) [reply lib/cylc/tests/test_review.py]: I had thought it was neater this way, since you have to cast tmp_path to a string, but I've changed it anyway.

**Review by wxtim — APPROVED** [review]

File: lib/cylc/review.py
Comment (oliver-sanders) [inline lib/cylc/review.py:266]: Now that you've switched over to the simpler non-db way of telling Cylc 8 from 7 I think we might be best to revert these DB closes for now in-case they have unintended side effects.

**Review by oliver-sanders — COMMENTED** [review]
Sorry, I keep getting distracted half-way through reviewing this.

LGTM but I think it would be safer to revert the db closes. Whilst I think it's the right thing to do, Cylc Review has worked for a long time without them. There's a risk we don't understand the interactions fully which I'd rather not take with this "final" Cylc 7 release :grin:!

File: lib/cylc/review.py
Comment (datamel) [reply lib/cylc/review.py]: I totally agree, I've removed them now.

**Review by oliver-sanders — APPROVED** [review]

**Comment by datamel** [comment]
@hjoliver, this pr covers the last lot of cylc review bugs that surfaced over the last week or so. We should be good to go for the next Cylc 7 release (when convenient).

**Comment by hjoliver** [comment]
> which I'd rather not take with this "final" Cylc 7 release :grin: !

Hey - no jokes about "final" releases! :sob:

**Pull request merged**
